src/Bicep.Core/TypeSystem/TypeAssignmentVisitor.cs

```diff
@@ -554,20 +568,32 @@ public override void VisitObjectPropertySyntax(ObjectPropertySyntax syntax)
 
                 if (syntax.Key is StringSyntax stringSyntax && stringSyntax.IsInterpolated())
                 {
-                    // if the key is an interpolated string, we need to check the expressions referenced by it 
+                    // if the key is an interpolated string, we need to check the expressions referenced by it
                     var keyType = typeManager.GetTypeInfo(syntax.Key);
                     CollectErrors(errors, keyType);
                     types.Add(keyType);
                 }
 
+                var vistingModuleParams = visitingModule
+                    && !allowedFunctionFlags.HasFlag(FunctionFlags.ModuleParamsAssignmentOnly)
+                    && syntax.Key is IdentifierSyntax identifierSyntax
+                    && string.Equals(identifierSyntax.IdentifierName, "params", StringComparison.OrdinalIgnoreCase);
+                if (vistingModuleParams)
+                {
+                    allowedFunctionFlags |= FunctionFlags.ModuleParamsAssignmentOnly;
+                }
                 var valueType = typeManager.GetTypeInfo(syntax.Value);
+                if (vistingModuleParams)
+                {
+                    allowedFunctionFlags &= ~FunctionFlags.ModuleParamsAssignmentOnly;
+                }
```

Comment:
This logic doesn't feel strictly related to type checking, and introducing new state into the type checker feels risky as the recursion is already very complicated. Would it be possible to implement this in the `EmitLimitationCalculator` class? Or is there anywhere else that this could fit?

Reply:
I remember struggling to place it elsewhere but a the end typeVisitor was the only place I had access to the information which function overload was actually used.

the best would be to put it in nameBindingVisitor where we check other flags, but that visitor does not visit instance Function calls. For instance functions, flags are checked here. Currently default is hardcoded, so there’s no flag checking on instance functions at all.

In fact we do not need this at all. Type assignment checking will do the work, as we can assign function result only to a string param with secure decorator. I initially planned to do so, but I added this upon review request.

Reply:
This check would only be needed because we have no way to emit other usages, so perfect fit for the EmitLimitation* classes (if we end up needing it after other conversations).

Reply:
Thing is that this fragment is not related to emitting. It’s for function flags which are used to limit functions available when assigning function result to a type.

Reply:
Function flags mechanism here, in TypeAssignmentVisitor could be useful in future if we’d like to have list* functions as instance functions, as they cannot be used in all places.

Reply:
As discussed, moved entire logic to a separate visitor